src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Constructors.jl

```diff
@@ -176,7 +176,7 @@ with coordinates [y1, y2, y3, y4, y5]
 ```
 """
 function affine_space(kk::BRT, n::Int; variable_name="x") where {BRT<:Ring}
-  R, _ = polynomial_ring(kk, [variable_name * "$i" for i in 1:n])
+  R, _ = polynomial_ring(kk, [variable_name * "$i" for i in 1:n]; cached=false)
```

Comment:
in the constructors for `affine_space` the ring should be cached!  If desired, we can add a keyword
`cached` and pass that on to `polynomial_ring`

Reply:
Why?

Reply:
For the same reason that `polynomial_ring` is cached by default.  (whether or not that reason is good)

Reply:
Cool, so you want the following behaviour?
```
julia> A1 = affine_space(NormalToricVariety, 3)
Normal toric variety

julia> A2 = affine_space(NormalToricVariety, 3)
Normal toric variety

julia> A1 == A2
false

julia> A1 = affine_space(QQ, 3)
Affine space of dimension 3
  over rational field
with coordinates [x1, x2, x3]

julia> A2 = affine_space(QQ, 3)
Affine space of dimension 3
  over rational field
with coordinates [x1, x2, x3]

julia> A1 == A2
true
```

Reply:
Good catch. In this case I would say that the `NormalToricVariety` has the unexpected behavior and should cache.

Reply:
(or possibly implement equality different)

Reply:
Currently rings are special as they basically only have one implementation. This is not true for varieties which exist as `AffineVariety`, `NormalToricVariety`, `Scheme`, etc. Your suggestion then needs to be passed on to every other implementation.

Reply:
Yes.

Reply:
> Yes.

So should caching then be extended to all standard constructions, like del pezzo, Hirzebruch, etc?

Reply:
The reason is that these affine spaces appear as keys in certain dictionaries. Sometimes we want different spaces.

Reply:
Ok, now you make a dictionary:
```
julia> A1 = affine_space(QQ, 3);

julia> A2 = affine_space(QQ, 3);

julia> d = Dict(A1=>1)
Dict{AffineVariety{QQField, QQMPolyRing}, Int64} with 1 entry:
  Affine 3-space over QQ with coordinates [x1, x2, x3] => 1

julia> d[A2]
ERROR: KeyError: key Affine 3-space over QQ with coordinates [x1, x2, x3] not found
Stacktrace:
 [1] getindex(h::Dict{AffineVariety{QQField, QQMPolyRing}, Int64}, key::AffineVariety{QQField, QQMPolyRing})
   @ Base ./dict.jl:498
 [2] top-level scope
   @ REPL[30]:1

julia> A1 == A2
true
```
I don't understand what you are doing.

Also you could make a `Dict{Scheme, FOO}` and then your proposed equality would cause this `Dict` to behave differently for different schemes.

Reply:
We  do IDdicts

Reply:
From my point of view this would be a breaking change of behavior.

Reply:
```
julia> idd = IdDict(A1=>1)
IdDict{AffineVariety{QQField, QQMPolyRing}, Int64} with 1 entry:
  Affine 3-space over QQ with coordinates [x1, x2, x3] => 1

julia> idd[A1]
1

julia> idd[A2]
ERROR: KeyError: key Affine 3-space over QQ with coordinates [x1, x2, x3] not found
Stacktrace:
 [1] getindex(d::IdDict{AffineVariety{QQField, QQMPolyRing}, Int64}, key::Any)
   @ Base ./iddict.jl:108
 [2] top-level scope
   @ REPL[35]:1
```

Reply:
> Currently rings are special as they basically only have one implementation. This is not true for varieties which exist as `AffineVariety`, `NormalToricVariety`, `Scheme`, etc. Your suggestion then needs to be passed on to every other implementation.

How does having several implementations relate to caching?
(Besides, we have many finite fields als well as far as I remember, we also have several implementations of the field of rational numbers etc.)
Do we have a rationale what is in the global cache and what not?

In any case I would be fine with caching `affine_space` as @fingolfin suggests. (but I don't care enough to implement it)

Reply:
> How does having several implementations relate to caching?

I illustrated this with the example `affine_space(NormalToricVariety, *)`, as caching would then have to be implemented for every new implementation of `affine_space` to ensure consistent behaviour. Opposed to only once if there is only one implementation.

> but I don't care enough to implement it

Ok, so who is going to write conformance tests for the caching mechanisms you want?